lib/auth/saml.go

```diff
@@ -443,3 +449,98 @@ func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, e
 
 	return re, nil
 }
+
+func getDecryptionCert(connector services.SAMLConnector) (*tls.Certificate, error) {
+	keypair := connector.GetEncryptionKeyPair()
+	if keypair == nil {
+		return nil, trace.Errorf("no encryption keypair configured in SAML connector")
+	}
+
+	decryptCert, err := tls.X509KeyPair([]byte(keypair.Cert), []byte(keypair.PrivateKey))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &decryptCert, nil
+}
+
+func RetrieveAssertionInfo(provider *saml2.SAMLServiceProvider, connector services.SAMLConnector, response *saml2types.Response, validate bool) (*saml2.AssertionInfo, error) {
+	assertionInfo := &saml2.AssertionInfo{
+		Values: make(saml2.Values),
+	}
+
+	var decryptCert *tls.Certificate
+	var err error
+	assertions := response.Assertions
+
+	if len(response.EncryptedAssertions) > 0 {
+		decryptCert, err = getDecryptionCert(connector)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	for _, encryptedAssertion := range response.EncryptedAssertions {
+		assertion, err := encryptedAssertion.Decrypt(decryptCert)
+		if err != nil {
+			return nil, trace.WrapWithMessage(err, "failed to decrypt encrypted SAML assertion")
+		}
+
+		assertions = append(assertions, *assertion)
+	}
+
+	if len(assertions) == 0 {
+		return nil, trace.Wrap(saml2.ErrMissingAssertion)
+	}
+
+	assertion := &assertions[0]
+	assertionInfo.Assertions = assertions
+	assertionInfo.ResponseSignatureValidated = response.SignatureValidated
+
+	var warningInfo *saml2.WarningInfo
+
+	if validate {
```

Comment:
Instead of passing the flag here, have the caller generate warning info if necessary.

Reply:
adjusted